### ISSUE TITLE
Lazy concatenation / block diagonal forms of operators

### ISSUE THREAD

**Issue by gaurav-arya** [SciML/SciMLOperators.jl#161]
Does there exist functionality analogous to those listed in https://julialinearalgebra.github.io/LinearMaps.jl/stable/types/#BlockMap-and-BlockDiagonalMap?

**Comment by vpuri3** [comment]
No, we don't have that functionality yet. It is on my to-do list. Would make writing vector-calculus operators really convenient. And will be very useful for solving coupled PDE systems.

**Comment by vpuri3** [comment]
Potential API:

```julia
struct BlockOperator
  Mblocks::Integer
  Nblocks::Integer
  Blocks

  function BlockOperator(Mblocks::Integer, Mblocks::Integer, blocks)
    @assert Mblocks * Nblocks == length(blocks)

    BlockOperator(Mblocks, Nblocks, blocks)
end

end

function BlockOperator(A::AbstractVecOrMat{<:AbstractSciMLOperator})
  BlockOperator(size(A), vec(A))
end

L = BlockOperator([A1, A2])
L * u = vcat(Tuple(B * u for B in L.blocks)...)
```

**Comment by gaurav-arya** [comment]
if we support sparse block patterns, then block diagonal becomes a special case

**Comment by vpuri3** [comment]
Yeah for sparse patterns, we can have a constructor that accepts pairs `(L, (m, n))`. And a special constructor for `BlockDiagonal`.

**Comment by vpuri3** [comment]
It might make sense to implement `BlockDiagonal` as its own class so we can define `has_ldiv`, `has_ldiv!` on it when each block is invertible.

**Comment by gaurav-arya** [comment]
Can always do `const BlockDiagonal = BlockOperator{AA} where {AA<:Diagonal}`, with `AA` imagined to be the block pattern

**Comment by gaurav-arya** [comment]
Also want to be mindful of code dup with kronecker products. $I \otimes A$ is a block diagonal matrix, and $A \otimes B = (A \otimes I)(B \otimes I)$, although that might not be the best perspective for efficiency

**Comment by vpuri3** [comment]
A case like `BlockDiagonal(A1, A2)` can't be covered by $I \otimes A$. But yes, there are many different expressions to write the same result in algebra

**Comment by vpuri3** [comment]
I remember putting in short-circuits in `TensorProductOperator` for when the outer is `IdentityOperator`. But we can always just write an overload if our lazy block version is faster.

**Comment by vpuri3** [comment]
`BlockOperators` also need to support lazy matrix algebra. So we can multiply blocks and compose their components. Example
```julia
G = BlockOperator([Dx, Dy])
T = DiagonalBlock([Tx, Ty])
Lapl = G' * T * G  # Dx' * Tx * Dx + Dy' * Ty * Dy
```

**Comment by vpuri3** [comment]
linking old block matrix issue here for completeness. https://github.com/SciML/SciMLOperators.jl/issues/45
